File: Phase-03/AI-Chatbot-Fullstack/backend/app/schemas/todo.py
"""
Todo request/response schemas.

Task Reference: T003 - Create TodoCreate, TodoUpdate, TodoResponse schemas (003-todo-crud)
Task Reference: T010 - Extend TodoCreate schema with due_date, priority, tag_ids
Task Reference: T011 - Extend TodoUpdate schema with due_date, priority, tag_ids
Task Reference: T012 - Extend TodoResponse schema to include tags array
Task Reference: T030 - Extend schemas for recurrence (006-recurring-reminders)
Task Reference: T041 - Extend TodoResponse for recurrence fields (006-recurring-reminders)
Feature: 003-todo-crud, 005-todo-enhancements, 006-recurring-reminders

Provides Pydantic schemas for:
- TodoCreate: Request body for creating a new todo
- TodoUpdate: Request body for partially updating a todo
- TodoResponse: Response body for todo data
- TagInTodoResponse: Embedded tag data in todo responses
"""

from datetime import date, datetime
from typing import List, Optional

from pydantic import BaseModel, Field, field_validator

from app.models.priority import Priority
from app.schemas.recurrence import RecurrenceConfig


class TagInTodoResponse(BaseModel):
    """
    Embedded tag data in todo responses.

    A lighter version of TagResponse without todo_count.
    Task Reference: T012
    """

    id: str = Field(..., description="Unique tag identifier")
    name: str = Field(..., description="Tag display name")
    color: str = Field(..., description="Hex color code")

    model_config = {"from_attributes": True}


class TodoCreate(BaseModel):
    """
    Request schema for creating a new todo.

    Task Reference: T011 - Add title validation (003-todo-crud)
    Task Reference: T010 - Add due_date, priority, tag_ids (005-todo-enhancements)
    """

    title: str = Field(
        ...,
        min_length=1,
        max_length=255,
        description="Task title (required, 1-255 chars)",
        examples=["Buy groceries"],
    )
    description: Optional[str] = Field(
        default=None,
        description="Optional detailed description",
        examples=["Milk, eggs, bread, cheese"],
    )
    # NEW fields for 005-todo-enhancements
    due_date: Optional[datetime] = Field(
        default=None,
        description="Optional deadline date and time (ISO 8601 format)",
        examples=["2026-01-25T15:00:00Z"],
    )
    priority: Optional[Priority] = Field(
        default=None,
        description="Optional priority level (high/medium/low)",
        examples=["high"],
    )
    tag_ids: Optional[List[str]] = Field(
        default=None,
        description="Optional list of tag IDs to assign",
        examples=[["tag_abc123", "tag_def456"]],
    )
    # NEW fields for 006-recurring-reminders
    recurrence: Optional[RecurrenceConfig] = Field(
        default=None,
        description="Optional recurrence configuration for repeating todos",
    )

    @field_validator("title")
    @classmethod
    def title_not_whitespace_only(cls, v: str) -> str:
        """Ensure title is not just whitespace."""
        if v.strip() == "":
            raise ValueError("Title cannot be empty or whitespace only")
        return v.strip()


class TodoUpdate(BaseModel):
    """
    Request schema for partially updating a todo.

    All fields are optional - only provided fields are updated.
    Task Reference: T011 - Add due_date, priority, tag_ids (005-todo-enhancements)
    """

    title: Optional[str] = Field(
        default=None,
        min_length=1,
        max_length=255,
        description="Updated title",
        examples=["Buy groceries and supplies"],
    )
    description: Optional[str] = Field(
        default=None,
        description="Updated description",
        examples=["Milk, eggs, bread, cheese, soap"],
    )
    completed: Optional[bool] = Field(
        default=None,
        description="Updated completion status",
        examples=[True],
    )
    # NEW fields for 005-todo-enhancements
    due_date: Optional[datetime] = Field(
        default=None,
        description="Updated deadline date and time (ISO 8601 format, or null to clear)",
        examples=["2026-01-25T15:00:00Z"],
    )
    priority: Optional[Priority] = Field(
        default=None,
        description="Updated priority level (high/medium/low, or null to clear)",
        examples=["medium"],
    )
    tag_ids: Optional[List[str]] = Field(
        default=None,
        description="Updated list of tag IDs (replaces existing tags)",
        examples=[["tag_abc123"]],
    )
    # NEW fields for 006-recurring-reminders
    recurrence: Optional[RecurrenceConfig] = Field(
        default=None,
        description="Updated recurrence configuration (or null to remove)",
    )

    @field_validator("title")
    @classmethod
    def title_not_whitespace_only(cls, v: Optional[str]) -> Optional[str]:
        """Ensure title is not just whitespace if provided."""
        if v is not None:
            if v.strip() == "":
                raise ValueError("Title cannot be empty or whitespace only")
            return v.strip()
        return v


class TodoResponse(BaseModel):
    """
    Response schema for todo data.

    Returns all todo fields including timestamps and tags.
    Task Reference: T012 - Add tags array (005-todo-enhancements)
    """

    id: str = Field(
        ...,
        description="Unique todo identifier",
        examples=["V1StGXR8_Z5jdHi6B-myT"],
    )
    title: str = Field(
        ...,
        description="Task title",
        examples=["Buy groceries"],
    )
    description: Optional[str] = Field(
        default=None,
        description="Task description",
        examples=["Milk, eggs, bread"],
    )
    completed: bool = Field(
        ...,
        description="Whether the task is completed",
        examples=[False],
    )
    user_id: str = Field(
        ...,
        description="Owner's user ID",
        examples=["user_abc123"],
    )
    created_at: datetime = Field(
        ...,
        description="Creation timestamp (UTC)",
        examples=["2026-01-17T10:30:00Z"],
    )
    updated_at: datetime = Field(
        ...,
        description="Last update timestamp (UTC)",
        examples=["2026-01-17T10:30:00Z"],
    )
    # NEW fields for 005-todo-enhancements
    due_date: Optional[datetime] = Field(
        default=None,
        description="Deadline date and time (ISO 8601 format)",
        examples=["2026-01-25T15:00:00Z"],
    )
    priority: Optional[Priority] = Field(
        default=None,
        description="Priority level (high/medium/low)",
        examples=["high"],
    )
    tags: List[TagInTodoResponse] = Field(
        default_factory=list,
        description="Tags assigned to this todo",
    )
    # NEW fields for 006-recurring-reminders
    is_recurring: bool = Field(
        default=False,
        description="Whether this todo recurs",
    )
    rrule: Optional[str] = Field(
        default=None,
        description="RFC 5545 RRULE string for recurrence pattern",
        examples=["FREQ=WEEKLY;BYDAY=MO,WE,FR"],
    )
    recurrence_end_date: Optional[date] = Field(
        default=None,
        description="Date when recurrence ends (if set)",
    )
    next_occurrence_date: Optional[date] = Field(
        default=None,
        description="Date of the next pending occurrence",
    )
    # T076, T078 [US4] - Current occurrence info for frontend toggle handling
    current_occurrence_id: Optional[str] = Field(
        default=None,
        description="ID of the current/next pending occurrence",
    )
    current_occurrence_status: Optional[str] = Field(
        default=None,
        description="Status of the current occurrence (pending/completed/skipped)",
    )

    model_config = {"from_attributes": True}

File: Phase-04/Kubernetes-Deployment/frontend/src/types/todo.ts
/**
 * Todo type definitions.
 *
 * Task Reference: T003 - Create Todo types per data-model.md (004-frontend-todo-ui)
 * Task Reference: T014 - Extend Todo type with due_date, priority, tags (005-todo-enhancements)
 * Task Reference: T046 - Extend Todo type for recurrence (006-recurring-reminders)
 * Feature: 004-frontend-todo-ui, 005-todo-enhancements, 006-recurring-reminders
 */

import type { Tag } from "./tag";
import type { RecurrenceConfig } from "./recurrence";

/**
 * Priority levels for todos.
 * Matches backend Priority enum.
 */
export type Priority = "high" | "medium" | "low";

/**
 * Priority display configuration.
 */
export const PRIORITY_CONFIG: Record<
  Priority,
  { label: string; color: string; bgColor: string }
> = {
  high: {
    label: "High",
    color: "#EF4444", // red-500
    bgColor: "#FEE2E2", // red-100
  },
  medium: {
    label: "Medium",
    color: "#EAB308", // yellow-500
    bgColor: "#FEF3C7", // yellow-100
  },
  low: {
    label: "Low",
    color: "#6B7280", // gray-500
    bgColor: "#F3F4F6", // gray-100
  },
};

/**
 * Todo item as returned by the backend API.
 */
export interface Todo {
  /** Unique identifier (nanoid, 21 chars) */
  id: string;

  /** Task title (1-255 chars, required) */
  title: string;

  /** Optional detailed description */
  description: string | null;

  /** Completion status */
  completed: boolean;

  /** Owner's user ID */
  user_id: string;

  /** Creation timestamp (ISO 8601, UTC) */
  created_at: string;

  /** Last update timestamp (ISO 8601, UTC) */
  updated_at: string;

  // NEW fields for 005-todo-enhancements
  /** Optional deadline date and time (ISO 8601 format) */
  due_date: string | null;

  /** Optional priority level */
  priority: Priority | null;

  /** Tags assigned to this todo */
  tags: Tag[];

  // NEW fields for 006-recurring-reminders
  /** Whether this todo recurs */
  is_recurring: boolean;

  /** RFC 5545 RRULE string for recurrence pattern */
  rrule: string | null;

  /** Date when recurrence ends (if set) */
  recurrence_end_date: string | null;

  /** Date of the next pending occurrence */
  next_occurrence_date: string | null;

  /** ID of the current/next pending occurrence (for T077 [US4]) */
  current_occurrence_id: string | null;

  /** Status of the current occurrence (for T078 [US4]) */
  current_occurrence_status: "pending" | "completed" | "skipped" | null;
}

/**
 * Request payload for POST /todos
 */
export interface TodoCreate {
  /** Task title (required, 1-255 chars) */
  title: string;

  /** Optional detailed description */
  description?: string;

  // NEW fields for 005-todo-enhancements
  /** Optional deadline date and time (ISO 8601 format) */
  due_date?: string | null;

  /** Optional priority level */
  priority?: Priority | null;

  /** Optional list of tag IDs to assign */
  tag_ids?: string[];

  // NEW fields for 006-recurring-reminders
  /** Optional recurrence configuration */
  recurrence?: RecurrenceConfig | null;
}

/**
 * Request payload for PATCH /todos/{id}
 * All fields are optional.
 */
export interface TodoUpdate {
  /** Updated title */
  title?: string;

  /** Updated description */
  description?: string;

  /** Updated completion status */
  completed?: boolean;

  // NEW fields for 005-todo-enhancements
  /** Updated deadline date and time (ISO 8601 format, or null to clear) */
  due_date?: string | null;

  /** Updated priority level (or null to clear) */
  priority?: Priority | null;

  /** Updated list of tag IDs (replaces existing tags) */
  tag_ids?: string[];
}

/**
 * State managed by useTodos hook
 */
export interface TodosState {
  /** List of todos */
  todos: Todo[];

  /** Loading state for initial fetch */
  isLoading: boolean;

  /** Loading state for mutations */
  isMutating: boolean;

  /** Error message if last operation failed */
  error: string | null;
}

/**
 * Tracks which todo is being edited
 */
export interface EditingState {
  /** ID of todo being edited, null if not editing */
  todoId: string | null;

  /** Form values for editing */
  formValues: {
    title: string;
    description: string;
    due_date: string | null;
    priority: Priority | null;
    tag_ids: string[];
  } | null;
}

/**
 * Toast notification type
 */
export type ToastType = "success" | "error" | "info";

/**
 * Toast notification data
 */
export interface Toast {
  id: string;
  message: string;
  type: ToastType;
}
